### PR TITLE
[WIP] Function ensemble

### LINKED ISSUE #1176: Create function ensemble
Create function ensemble that can evaluate realizations running only one function, that also integrates with the ensemble evaluator and produce events

### PR DESCRIPTION
**Pull request by sondreso**
**Issue**
Resolves #1176 

See [showcase_function_ensemble.py](https://github.com/equinor/ert/blob/f91ba4fcdcd6b7c16a13724c03538b085fa0d792/showcase_function_ensemble.py) for a demo!

### CODE REVIEW

File: ert_shared/ensemble_evaluator/entity/ensemble.py [hunk 1]
```diff
@@ -125,10 +143,33 @@ def build(self):
...
+class _FunctionJobBuilder(_BaseJobBuilder):
+    def __init__(self):
+        super().__init__()
+        self._fun = None
+        self.reset()
+
+    def reset(self):
+        super().reset()
+        self._fun = None
+        return self
```
Comment (lars-petter-hauge) [inline ert_shared/ensemble_evaluator/entity/ensemble.py:155]: Could I ask the reason for returning self in these functions?

Comment (sondreso) [reply ert_shared/ensemble_evaluator/entity/ensemble.py:155]: It allows for a builder pattern, so you can do:

`create_ensemble_builder(...).add_job(...).do_some_stuff(...).reset(...).do_some_stuff(..).build(...)`

File: showcase_function_ensemble.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    ee.run()
+    ensemble.join()
+    ee.stop()
+
+    return ensemble.results()
+
+
+def main():
+    inputs = generate_coefficients()
+    results = evaluate(inputs, fun=polynomial)
```
Comment (markusdregi) [inline showcase_function_ensemble.py:59]: This is really neat 👏

File: ert_shared/ensemble_evaluator/entity/ensemble.py (hunk 1 above)
Comment (jondequinor) [reply ert_shared/ensemble_evaluator/entity/ensemble.py:155]: https://refactoring.guru/design-patterns/builder

File: ert_shared/ensemble_evaluator/config.py [hunk 3]
```diff
@@ -6,7 +6,7 @@
 CLIENT_URI = "client"
 DISPATCH_URI = "dispatch"
 DEFAULT_PORT = "8765"
-DEFAULT_HOST = "localhost"
+DEFAULT_HOST = "0.0.0.0"
```
Comment (jondequinor) [inline ert_shared/ensemble_evaluator/config.py:9]: I think this should be exceptional behavior? `0.0.0.0` is maybe a bit too open, especially without any auth.

Comment (sondreso) [reply ert_shared/ensemble_evaluator/config.py:9]: Agreed, this was just to do some testing and should be removed before we merge it!

**Pull request closed without merging**